tensorpipe/channel/cuda_basic/channel.cc

```diff
@@ -0,0 +1,358 @@
+  uint64_t nextTensorBeingSent_{0};
+
+  // Increasing identifier for recv operations.
+  uint64_t nextTensorBeingReceived_{0};
+
+  // An identifier for the channel, composed of the identifier for the context,
+  // combined with an increasing sequence number. It will only be used for
+  // logging and debugging purposes.
+  std::string id_;
+
+  EagerCallbackWrapper<Impl> eagerCallbackWrapper_{*this, this->loop_};
+
+  // For some odd reason it seems we need to use a qualified name here...
+  template <typename T>
+  friend class tensorpipe::EagerCallbackWrapper;
+};
+
+Channel::Channel(
+    ConstructorToken /* unused */,
+    std::shared_ptr<Context::PrivateIface> context,
+    std::shared_ptr<CpuChannel> cpuChannel,
+    std::shared_ptr<CudaLoop> cudaLoop,
+    std::string id)
+    : impl_(std::make_shared<Impl>(
+          std::move(context),
+          std::move(cpuChannel),
+          std::move(cudaLoop),
+          std::move(id))) {
+  impl_->init();
+}
+
+Channel::Impl::Impl(
+    std::shared_ptr<Context::PrivateIface> context,
+    std::shared_ptr<CpuChannel> cpuChannel,
+    std::shared_ptr<CudaLoop> cudaLoop,
+    std::string id)
+    : context_(std::move(context)),
+      cpuChannel_(std::move(cpuChannel)),
+      cudaLoop_(std::move(cudaLoop)),
+      closingReceiver_(context_, context_->getClosingEmitter()),
+      id_(std::move(id)) {}
+
+void Channel::Impl::init() {
+  loop_.deferToLoop([this]() { initFromLoop(); });
+}
+
+void Channel::Impl::initFromLoop() {
+  TP_DCHECK(loop_.inLoop());
+  closingReceiver_.activate(*this);
+}
+
+void Channel::send(
+    CudaBuffer buffer,
+    TDescriptorCallback descriptorCallback,
+    TSendCallback callback) {
+  impl_->send(buffer, std::move(descriptorCallback), std::move(callback));
+}
+
+void Channel::Impl::send(
+    CudaBuffer buffer,
+    TDescriptorCallback descriptorCallback,
+    TSendCallback callback) {
+  loop_.deferToLoop([this,
+                     buffer,
+                     descriptorCallback{std::move(descriptorCallback)},
+                     callback{std::move(callback)}]() mutable {
+    sendFromLoop(buffer, std::move(descriptorCallback), std::move(callback));
+  });
+}
+
+// Send memory region to peer.
+void Channel::Impl::sendFromLoop(
+    CudaBuffer buffer,
+    TDescriptorCallback descriptorCallback,
+    TSendCallback callback) {
+  TP_DCHECK(loop_.inLoop());
+
+  const uint64_t sequenceNumber = nextTensorBeingSent_++;
+  TP_VLOG(4) << "Channel " << id_ << " received a send request (#"
+             << sequenceNumber << ")";
+
+  if (error_) {
+    descriptorCallback(error_, std::string());
+    callback(error_);
+    return;
+  }
+
+  TP_VLOG(4) << "Channel " << id_
+             << " is copying buffer from CUDA device to CPU";
+  auto tmpBuffer = makeCudaPinnedBuffer(buffer.length);
+  TP_CUDA_CHECK(cudaMemcpyAsync(
+      tmpBuffer.get(),
+      buffer.ptr,
+      buffer.length,
+      cudaMemcpyDeviceToHost,
+      buffer.stream));
+
+  auto cudaCallback =
```

Comment:
Shouldn't this be wrapped inside eagerCallbackWrapper? As it is, it only captures a _raw pointer_ to `this`, with no other means of keeping the Channel::Impl alive until the callback has fired. Which means the Channel::Impl could be destroyed (if the user closes it) and then this callback would cause a segfault when dereferencing `this`.

Reply:
Oh, never mind, I saw you added a TODO for this just a few lines down.